streams/src/test/java/org/apache/kafka/streams/integration/VersionedKeyValueStoreIntegrationTest.java

```diff
@@ -334,18 +399,65 @@ private final int produceSourceData(final long timestamp,
     }
 
     /**
-     * Test-only processor for inserting records into a versioned store while also tracking
-     * them separately in-memory, and performing checks to validate expected store contents.
-     * Forwards the number of failed checks downstream for consumption.
+     * @param topic       topic to produce to
+     * @param dataTracker map of key -> timestamp -> value for tracking data which is produced to
+     *                    the topic. This method will add the produced data into this in-memory
+     *                    tracker in addition to producing to the topic, in order to keep the two
+     *                    in sync.
+     * @param timestamp   timestamp to produce with
+     * @param keyValues   key-value pairs to produce
+     *
+     * @return number of records produced
+     */
+    @SuppressWarnings("varargs")
+    @SafeVarargs
+    private final int produceDataToTopic(final String topic,
+                                         final DataTracker dataTracker,
+                                         final long timestamp,
+                                         final KeyValue<Integer, String>... keyValues) {
+        produceDataToTopic(topic, timestamp, keyValues);
+
+        for (final KeyValue<Integer, String> keyValue : keyValues) {
+            dataTracker.add(keyValue.key, timestamp, keyValue.value);
+        }
+
+        return keyValues.length;
+    }
+
+    /**
+     * Test-only processor for validating expected contents of a versioned store, and forwards
+     * the number of failed checks downstream for consumption. Callers specify whether the
+     * processor should also be responsible for inserting records into the store (while also
+     * tracking them separately in-memory for use in validation).
      */
     private static class VersionedStoreContentCheckerProcessor implements Processor<Integer, String, Integer, Integer> {
 
         private ProcessorContext<Integer, Integer> context;
         private VersionedKeyValueStore<Integer, String> store;
 
+        // whether or not the processor should write records to the store as they arrive.
+        // must be false for global stores.
```

Comment:
Why must it be false for global stores?

Reply:
When a processor accesses a global store, the store is read-only: https://github.com/apache/kafka/blob/404a833df7de9a7d5efe35beb5bafb7c6972601e/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java#L167-L170

Reply:
Know that I understand how the test actually works, it makes sense.